dot/core/service.go

```diff
@@ -350,11 +350,29 @@ func (s *Service) ProcessBlockAnnounceMessage(msg network.Message) error {
 		return errors.New("could not cast network.Message to BlockAnnounceMessage")
 	}
 
-	if s.blockState == nil {
-		return errors.New("ProcessBlockAnnounceMessage error: blockState is nil")
+	header, err := types.NewHeader(blockAnnounceMessage.ParentHash, blockAnnounceMessage.Number, blockAnnounceMessage.StateRoot, blockAnnounceMessage.ExtrinsicsRoot, blockAnnounceMessage.Digest)
+	if err != nil {
+		return err
 	}
 
-	latestBlockNum := s.blockState.LatestHeader().Number
+	_, err = s.blockState.GetHeader(header.Hash())
+	if err != nil && err.Error() == "Key not found" {
+		err = s.blockState.SetHeader(header)
+		if err != nil {
+			return err
+		}
+	} else {
+		return err
+	}
+
+	log.Info("[core] imported block", "number", header.Number, "hash", header.Hash())
```

Comment:
We only imported block if we set the header, correct? Should this be in the same block of code as `SetHeader`?

Reply:
good point, will update!